src/lib.rs

```diff
@@ -1248,26 +1230,23 @@ pub unsafe extern "C" fn dav1d_data_wrap(
     result.into()
 }
 
-pub(crate) unsafe fn rav1d_data_wrap_user_data(
-    buf: *mut Rav1dData,
-    user_data: *const u8,
-    free_callback: Option<unsafe extern "C" fn(*const u8, *mut c_void) -> ()>,
-    cookie: *mut c_void,
-) -> Rav1dResult {
-    return rav1d_data_wrap_user_data_internal(buf, user_data, free_callback, cookie);
-}
-
 #[no_mangle]
 pub unsafe extern "C" fn dav1d_data_wrap_user_data(
     buf: *mut Dav1dData,
     user_data: *const u8,
     free_callback: Option<unsafe extern "C" fn(*const u8, *mut c_void) -> ()>,
     cookie: *mut c_void,
 ) -> Dav1dResult {
-    let mut buf_rust = buf.read().into();
-    let result = rav1d_data_wrap_user_data(&mut buf_rust, user_data, free_callback, cookie);
-    buf.write(buf_rust.into());
-    result.into()
+    || -> Rav1dResult {
+        let buf = validate_input!(NonNull::new(buf).ok_or(EINVAL))?;
+        // Note that `dav1d` doesn't do this check, but they do for the similar [`dav1d_data_wrap`].
+        let user_data = validate_input!(NonNull::new(user_data.cast_mut()).ok_or(EINVAL))?;
+        let mut data = buf.as_ptr().read().to::<Rav1dData>();
+        data.wrap_user_data(user_data, free_callback, cookie)?;
+        buf.as_ptr().write(data.into());
+        Ok(())
+    }()
```

Comment:
im begging, screaming, crying for a `try` block to save us from this javascript-ass shit

Reply:
(this is not a criticism, just a jumping-off point for future discussion)